### PR TITLE
fix(tidy3d): FXC-5005-web-estimate-cost-printing-duplicate-output

### PR DESCRIPTION
**Pull request by marcorudolphflex**
> [!NOTE]
> Eliminates duplicate cost estimation output when uploading a `Job` by separating implicit vs. explicit verbosity and plumbing a dedicated flag.
> 
> - Suppresses `estimate_cost` during implicit `Job.task_id` resolution by calling `_upload(verbose_estimate_cost=False)`; `upload()` now explicitly calls `estimate_cost(verbose=True)` to print once
> - Adds `verbose_estimate_cost` to `web.upload` and threads it through `Job._upload`
> - Extends test fakes to honor verbosity and adds `test_job_estimate_cost_logging` to assert correct print counts
> - Updates `CHANGELOG.md` to note the fix
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 0eeacefb01a8bef07c3bc602d71544fbd9ba28c1. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

<h3>Greptile Summary</h3>

This PR fixes duplicate cost estimation output when uploading a `Job`. The issue occurred because accessing `job.task_id` (which implicitly uploads) and calling `job.upload()` both triggered cost estimation printing.

**Key Changes:**
- Modified `Job.task_id` property to call `_upload(verbose=False)` instead of `_upload()`, suppressing the implicit cost estimation print
- Updated `_upload()` method to accept optional `verbose` parameter and forward it to `web.upload()`
- Modified `Job.upload()` to explicitly call `estimate_cost(verbose=True)` before accessing `task_id`, ensuring single controlled output
- Enhanced test mocks (`fake_estimate_cost` and `fake_upload`) to respect the `verbose` parameter
- Added comprehensive test `test_job_estimate_cost_logging` that verifies cost estimation prints exactly once during upload and respects verbose flags

The fix elegantly separates concerns: implicit operations (accessing `task_id`) run silently, while explicit operations (`upload()`) provide user feedback when appropriate.

<h3>Confidence Score: 5/5</h3>

- This PR is safe to merge with minimal risk
- The changes are well-contained and surgical, addressing a specific UX issue without affecting core logic. The implementation correctly uses the existing verbose parameter pattern, and comprehensive tests verify the behavior. No breaking changes or edge cases were identified.
- No files require special attention

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| tidy3d/web/api/container.py | Modified task_id property and _upload/_upload methods to prevent duplicate cost estimation printing by suppressing verbose output when accessing task_id property |
| tests/test_web/test_webapi.py | Added verbose-aware fakes and new test test_job_estimate_cost_logging to verify cost estimation is printed exactly once during upload |
| CHANGELOG.md | Added changelog entry for duplicate cost estimation fix under Fixed section |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant User
    participant Job
    participant task_id as task_id (property)
    participant _upload
    participant web.upload
    participant web.estimate_cost
    
    Note over User,web.estimate_cost: Before Fix: Duplicate Printing
    User->>Job: job.upload()
    Job->>task_id: access self.task_id
    task_id->>_upload: _upload()
    _upload->>web.upload: web.upload(verbose=True)
    web.upload->>web.estimate_cost: estimate_cost(verbose=True)
    web.estimate_cost-->>User: 🖨️ Print cost
    
    Note over User,web.estimate_cost: After Fix: Single Printing
    User->>Job: job.upload()
    Job->>Job: if self.verbose: estimate_cost(verbose=True)
    Job->>web.estimate_cost: web.estimate_cost(task_id, verbose=True)
    web.estimate_cost-->>User: 🖨️ Print cost (once)
    Job->>task_id: access self.task_id
    task_id->>_upload: _upload(verbose=False)
    _upload->>web.upload: web.upload(verbose=False)
    web.upload->>web.estimate_cost: estimate_cost(verbose=False)
    Note over web.estimate_cost: No print (verbose=False)
```

**Context used:**

- Rule from `dashboard` - Write changelog entries to be clear and descriptive from a user's perspective, focusing on the added... ([source](https://app.greptile.com/review/custom-context?memory=66e7d342-0019-4d74-a4fb-81567fea2f3c))

### CODE REVIEW

File: tidy3d/web/api/container.py [hunk 1]
```diff
@@ -360,19 +360,23 @@ def task_id(self) -> TaskId:
         if self.task_id_cached:
             return self.task_id_cached
         self._check_folder(self.folder_name)
-        return self._upload()
+        return self._upload(verbose=False)
 
-    def _upload(self) -> TaskId:
+    def _upload(self, verbose: Optional[bool] = None) -> TaskId:
```
Comment (momchil-flex) [inline tidy3d/web/api/container.py]: Why `Optional[bool] = None` and not just `bool=False`?

**Review by momchil-flex — APPROVED** [review]

File: tidy3d/web/api/container.py (hunk 1 above)
Comment (marcorudolphflex) [reply tidy3d/web/api/container.py]: should override `self.verbose` only if provided

**Comment by marcorudolphflex** [comment]
Just noticed that we do not have any verbosity on Job.upload now beside estimate cost, right? May reconsider this handling...

**Review by daquinteroflex — APPROVED** [review]

File: tidy3d/web/api/webapi.py [hunk 2]
```diff
@@ -561,7 +564,9 @@ def upload(
         remote_sim_file=remote_sim_file,
     )
 
-    estimate_cost(task_id=resource_id, solver_version=solver_version, verbose=verbose)
+    run_cost_estimate = verbose if run_cost_estimate is None else run_cost_estimate
+    if run_cost_estimate:
+        estimate_cost(task_id=resource_id, solver_version=solver_version, verbose=verbose)
```
Comment (daquinteroflex) [inline tidy3d/web/api/webapi.py]: So the issue with this is that if verbose is false, it should still run `estimate_cost` to trigger the metadata

Comment (marcorudolphflex) [reply tidy3d/web/api/webapi.py]: oh did not know that, will revert

**Review by daquinteroflex — CHANGES_REQUESTED** [review]
Does this enable running cost estimate by default when verbose is false in upload?

**Comment by marcorudolphflex** [comment]
> Does this enable running cost estimate by default when verbose is false in upload?

now it should do that

**Review by daquinteroflex — APPROVED** [review]
Thanks that'll do it!
